components/Parsing/__test__/TokenizerTest.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+use PHPDoc\Internal\Parsing\Exception\UnexpectedTokenException;
+use PHPDoc\Internal\Parsing\Token;
+use PHPDoc\Internal\Parsing\Tokenizer;
+use PHPDoc\Internal\Testing\Assert;
+
+class TokenizerTest extends Assert
+{
+    const STRING_REGEX = '^\"[^\".]*\"';
+    const WHITE_SPACE_REGEX = '^ +';
+    const CR_REGEX = '^\n';
+
+    public function testSrc()
+    {
+        $tokenizer = new Tokenizer('foo');
+        $this->assertSame('foo', $tokenizer->getSrc());
+
+        $tokenizer->setSrc('bar');
+        $this->assertSame('bar', $tokenizer->getSrc());
+    }
+
+    public function testHandleToken()
+    {
+        $tokenizer = new Tokenizer();
+        $token = new Token('bar', 'a', ['a' => 'b'], false);
+
+        $tokenizer->addCustomToken($token);
+        $tokenizer->addToken('foo', 'b', ['b' => 'c']);
+        $tokenizer->addSkipToken('baz', 'c', ['c' => 'd']);
+        $tokens = $tokenizer->getTokens();
+
+        $token = $tokens[0];
+        $this->assertCount(3, $tokens);
+        $this->assertInstanceOf($token, Token::class);
+        $this->assertSame('bar', $token->getType());
+        $this->assertSame('a', $token->getRegex());
+        $this->assertSame(['a' => 'b'], $token->getMetadata());
+        $this->assertFalse($token->toSkip());
+        $this->assertSame(1, $token->getStartLine());
+        $this->assertSame(1, $token->getEndLine());
+
+        $token = $tokens[1];
+        $this->assertInstanceOf($token, Token::class);
+        $this->assertSame('foo', $token->getType());
+        $this->assertSame('b', $token->getRegex());
+        $this->assertSame(['b' => 'c'], $token->getMetadata());
+        $this->assertFalse($token->toSkip());
+        $this->assertSame(1, $token->getStartLine());
+        $this->assertSame(1, $token->getEndLine());
+
+        $token = $tokens[2];
+        $this->assertInstanceOf($token, Token::class);
+        $this->assertSame('baz', $token->getType());
+        $this->assertSame('c', $token->getRegex());
+        $this->assertSame(['c' => 'd'], $token->getMetadata());
+        $this->assertTrue($token->toSkip());
+        $this->assertSame(1, $token->getStartLine());
+        $this->assertSame(1, $token->getEndLine());
+    }
+
+    public function testTokenize()
+    {
+        $tokenizer = new Tokenizer();
+        $tokenizer->setSrc($src = <<<'EOD'
+"1
+2
+3"
+EOD);
```

Comment:
```suggestion
        $tokenizer = new Tokenizer(<<<'EOD'
"1
2
3"
EOD);
```
Seems pointless to set it outside the constructor

Reply:
Je dois conserver la déclaration `$src` car je m'en sers plus bas:
```php
$this->assertSame($src, $token->getText());
```
Mais va pour:
```php
$tokenizer = new Tokenizer($src = <<<'EOD'
"1
2
3"
EOD);
```